### ISSUE TITLE
Messges Sent on Stream Are Sent By Ref Event Though They Are Not Marked as Immutable

### ISSUE THREAD

**Issue by zeus82** [dotnet/orleans#6637]
I have reproduced the problem in [this repo](https://github.com/zeus82/orleans_not_serializing).

Basically, the `HelloGrain` sends 3 messages on a stream, changing a value in an existing object each time.  The `HelloArchiveGrain` receives them and stores them in a list.  The `HelloArchiveGrain` prints the value of the first element in the list.  That value should never change, but when the next 2 messages are received, the value has changed.  This could only happen if the object was passed by ref.

This is the output:

```
The value of the first element is 1
The value of the first element is 5
The value of the first element is 50
```

The value of the first element should always be 1!!!

**Comment by benjaminpetit** [comment]
I don't see the same output in your repro, but did you try to configure `SimpleMessageStreamProviderOptions.OptimizeForImmutableData` to `false`? By default it is `true`, that means that it will assume that all data used in streaming are immutable, so it will not do a deep copy.

I agree the name is a bit confusing, and the default value might not be a safe one.

**Comment by zeus82** [comment]
Sorry, I think I missed a commit in the sample repo, but when I set 'SimpleMessageStreamProviderOptions.OptimizeForImmutableData' to false, I get the behaviour I expect.  I added a note about this to the documentation: https://github.com/dotnet/orleans/pull/6651

**Comment by sergeybykov** [comment]
Resolved via #6651.
